spec/ataru/virkailija/virkailija_routes_spec.clj

```diff
@@ -82,4 +101,93 @@
+                                         (create-element "B")
+                                         (create-element "C")))
+            form (:body resp)
+            with-update (-> form
+                            (update :content (fn [v] [(first v) (last v)])))
+            operations (form-diff/as-operations form with-update)
+            new-content (get-content-from-response (batch-form (:id form) operations))]
+        (should= (get-names new-content) ["A" "C"])))
+
+  (it "Should handle updates"
+      (let [resp (post-form (create-form (create-element "A")
+                                         (create-element "B")
+                                         (create-element "C")))
+            form (:body resp)
+            with-updates (-> form
+                             (update-in [:content 0 :label :fi] (fn [e] "AA"))
+                             (update-in [:content 1 :label :fi] (fn [e] "BB")))
+            operations (form-diff/as-operations form with-updates)
+            new-content (get-content-from-response (batch-form (:id form) operations))]
+        (should= (get-names new-content) ["AA" "BB" "C"])))
+
+  (it "Should handle relocation"
+      (let [resp (post-form (create-form (create-element "A")
+                                         (create-element "B")
+                                         (create-element "C")))
+            form (:body resp)
+            with-update (-> form
+                            (update :content (fn [v] (swap v 0 1))))
+            operations (form-diff/as-operations form with-update)
+            new-content (get-content-from-response (batch-form (:id form) operations))]
+        (should= (get-names new-content) ["B" "A" "C"])))
+
+  (it "Should handle move out of wrapper element"
+      (let [resp        (post-form (create-form (create-wrapper (create-element "A1") (create-element "A2"))
+                                                (create-element "B")
+                                                (create-element "C")))
+            form        (:body resp)
+            with-update (-> form
+                            (update :content (fn [content]
+                                                 (let [[wrapper & rest] content
+                                                       a1    (get-in content [0 :children 0])
+                                                       a2    (get-in content [0 :children 1])
+                                                       wa2   (assoc wrapper :children [a2])
+                                                       new-c (concat [a1 wa2] rest)]
+                                                   new-c))))
+            operations (form-diff/as-operations form with-update)
+            new-content (get-content-from-response (batch-form (:id form) (form-diff/as-operations form with-update)))]
+        (should= (get-structure-as-names new-content) ["A1" ["A2"] "B" "C"])))
+
+  )
+
```

Comment:
Pitäiskö tässä testata vielä semmosta "concurrent" muokkausta, eli pistää kaks batch-formia peräkkäin samalle alkuperäiselle formille eri muutoksilla?